core-ktx/src/main/java/com/amplifyframework/kotlin/facades/KotlinDataStoreFacade.kt

```diff
@@ -0,0 +1,161 @@
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.kotlin.facades
+
+import com.amplifyframework.core.Amplify
+import com.amplifyframework.core.async.Cancelable
+import com.amplifyframework.core.model.Model
+import com.amplifyframework.core.model.query.QueryOptions
+import com.amplifyframework.core.model.query.predicate.QueryPredicate
+import com.amplifyframework.datastore.DataStoreCategoryBehavior as Delegate
+import com.amplifyframework.datastore.DataStoreException
+import com.amplifyframework.datastore.DataStoreItemChange
+import com.amplifyframework.kotlin.DataStore
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.sendBlocking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+
+class KotlinDataStoreFacade(private val delegate: Delegate = Amplify.DataStore) : DataStore {
+    @Throws(DataStoreException::class)
+    override suspend fun <T : Model> save(item: T, predicate: QueryPredicate) {
+        return suspendCoroutine { continuation ->
+            delegate.save(
+                item,
+                predicate,
+                { continuation.resume(Unit) },
+                { continuation.resumeWithException(it) }
+            )
+        }
+    }
+
+    @Throws(DataStoreException::class)
+    override suspend fun <T : Model> delete(item: T, predicate: QueryPredicate) {
+        return suspendCoroutine { continuation ->
+            delegate.delete(
+                item,
+                predicate,
+                { continuation.resume(Unit) },
+                { continuation.resumeWithException(it) }
+            )
+        }
+    }
+
+    @ExperimentalCoroutinesApi
+    @Throws(DataStoreException::class)
+    override fun <T : Model> query(itemClass: Class<T>, options: QueryOptions): Flow<T> {
+        return callbackFlow {
+            delegate.query(
+                itemClass,
+                options,
+                {
+                    while (it.hasNext()) {
+                        sendBlocking(it.next())
+                    }
+                    close()
+                },
+                { close(it) }
+            )
+            awaitClose {}
+        }
+    }
+
+    @ExperimentalCoroutinesApi
+    @Throws(DataStoreException::class)
+    override fun observe(): Flow<DataStoreItemChange<out Model>> {
+        return callbackFlow {
+            val cancelable = AtomicReference<Cancelable?>()
+            delegate.observe(
+                { cancelable.set(it) },
```

Comment:
Ah, yea. Actually I just prototyped this, and __it is awesome__.

With Rx, It would have been like a `Single<Observable<DataStoreItemChange<T>>`, and it would be kind of ugly to deal with.

But since coroutines get rid of the await-style boiler plate, I implemented your solution and literally didn't have to change my unit tests. Everything was already getting launched in a coroutine scope and it just worked.

Reply:
Awesome!  Would a similar approach work for `API.subscribe` too?